### ISSUE TITLE
[Multi-device] Problems with argmax on sharded/replicated tensor

### ISSUE THREAD

**Issue by umalesTT** [tenstorrent/tt-mlir#3963]
### Describe the bug

When computing accuracy at the end of training for classification, argmax fails if we are in multi-device mode. Below is small repro:
StableHLO code:
```
module @jit_sharded_argmax attributes {mhlo.num_partitions = 2 : i32, mhlo.num_replicas = 1 : i32} {
  sdy.mesh @mesh = <["dp"=2]>
  func.func public @main(%arg0: tensor<256x10xf32> {sdy.sharding = #sdy.sharding<@mesh, [{}, {}]>}) -> (tensor<256xi32> {jax.result_info = "", sdy.sharding = #sdy.sharding<@mesh, [{}]>}) {
    %0 = sdy.manual_computation(%arg0) in_shardings=[<@mesh, [{}, {}]>] out_shardings=[<@mesh, [{}]>] manual_axes={"dp"} (%arg1: tensor<256x10xf32>) {
      %c = stablehlo.constant dense<0> : tensor<i32>
      %cst = stablehlo.constant dense<0xFF800000> : tensor<f32>
      %1 = stablehlo.iota dim = 0 : tensor<10xi32>
      %2 = stablehlo.broadcast_in_dim %1, dims = [1] : (tensor<10xi32>) -> tensor<256x10xi32>
      %3:2 = stablehlo.reduce(%arg1 init: %cst), (%2 init: %c) across dimensions = [1] : (tensor<256x10xf32>, tensor<256x10xi32>, tensor<f32>, tensor<i32>) -> (tensor<256xf32>, tensor<256xi32>)
       reducer(%arg2: tensor<f32>, %arg4: tensor<f32>) (%arg3: tensor<i32>, %arg5: tensor<i32>)  {
        %4 = stablehlo.compare  GT, %arg2, %arg4,  FLOAT : (tensor<f32>, tensor<f32>) -> tensor<i1>
        %5 = stablehlo.compare  NE, %arg2, %arg2,  FLOAT : (tensor<f32>, tensor<f32>) -> tensor<i1>
        %6 = stablehlo.or %4, %5 : tensor<i1>
        %7 = stablehlo.compare  EQ, %arg2, %arg4,  FLOAT : (tensor<f32>, tensor<f32>) -> tensor<i1>
        %8 = stablehlo.compare  LT, %arg3, %arg5,  SIGNED : (tensor<i32>, tensor<i32>) -> tensor<i1>
        %9 = stablehlo.and %7, %8 : tensor<i1>
        %10 = stablehlo.or %6, %9 : tensor<i1>
        %11 = stablehlo.select %6, %arg2, %arg4 : tensor<i1>, tensor<f32>
        %12 = stablehlo.select %10, %arg3, %arg5 : tensor<i1>, tensor<i32>
        stablehlo.return %11, %12 : tensor<f32>, tensor<i32>
      }
      sdy.return %3#1 : tensor<256xi32>
    } : (tensor<256x10xf32>) -> tensor<256xi32>
    return %0 : tensor<256xi32>
  }
}
```

fails with 

```
error: failed to legalize unresolved materialization from () to ('tensor<256xi32, #ttcore.mesh_sharding<"mesh">>') that remained live after conversion
```

I will circle back to debugging this later.
